File: bank/lib/bank.rb
require 'virtus'
require 'set'

# Public: Handle clients in a Bank.
#
# Examples
#
#   clients = [
#     Bank::Client.new(arrival: 1,  service_duration: 5 ),
#     Bank::Client.new(arrival: 0,  service_duration: 10),
#     Bank::Client.new(arrival: 0,  service_duration: 10),
#     Bank::Client.new(arrival: 1,  service_duration: 10),
#     Bank::Client.new(arrival: 2,  service_duration: 10),
#     Bank::Client.new(arrival: 30, service_duration: 10),
#   ]
#
#   Bank.new
#       .open(1)
#       .add_clients_to_line(clients)
#       .close
#       .clients_that_waited_more_than(20)
#   # => 1
class Bank

  # Public: Error raised when invalid operations are tried with the Bank.
  # For instance, trying to open an already opened Bank, of trying to close
  # an already closed Bank.
  class InvalidStateError < StandardError
  end

  # Public: A Client in the line of the Bank.
  class Client
    include Virtus

    # Public: The Integer minute the Client arrived
    # at the bank.
    attribute :arrival, Integer

    # Public: The Integer amount of minutes it takes
    # to serve the Client.
    attribute :service_duration, Integer

    # Public: The Integer amount of minutes the Client
    # had to wait in line.
    attribute :waiting_duration, Integer, default: 0

    # Public: The Integer amount of minutes a teller
    # has been serving the Client.
    attribute :serving_duration, Integer, default: 0

    # Public: The Symbol state in which the Client is. Allowed states are
    # :before_line, :in_line, :being_served and :served.
    attribute :state, Symbol, default: :before_line

    def before_line?
      state == :before_line
    end

    def in_line?
      state == :in_line
    end

    def being_served?
      state == :being_served
    end

    def served?
      state == :served
    end

    def finished_serving?
      being_served? && service_duration == serving_duration
    end
  end

  # Public: Returns the Integer number of tellers.
  attr_reader :tellers

  # Public: Returns the Symbol current state of the Bank. Allowed states are
  # :open and :closed.
  attr_reader :state

  # Public: Returns the Array of Clients in the Bank.
  attr_reader :clients

  # Public: Initialize a Bank.
  def initialize
    @state   = :closed
    @tellers = 0
    @clients = Array.new
  end

  # Public: Open the Bank for the day.
  #
  # tellers - The Integer number of tellers in the Bank for the day.
  #
  # Returns the Bank itself, useful for chaining.
  # Raises Bank::InvalidStateError if the Bank is already open.
  def open tellers
    if open?
      raise InvalidStateError.new 'You tried to open an already opened Bank'
    end

    @state   = :open
    @tellers = tellers
    clients.clear
    self
  end

  # Public: Tell if the Bank is open.
  #
  # Returns the Boolean indicating if the Bank is open.
  def open?
    state == :open
  end

  # Public: Close the bank.
  #
  # Returns the Bank itself, useful for chaining.
  # Raises Bank::InvalidStateError if the Bank is already closed.
  def close
    unless open?
      raise InvalidStateError.new 'You tried to close an already closed Bank'
    end

    @state = :closed
    calculate_wating_times unless clients.empty?
    self
  end

  # Public: Add Clients to line in Bank.
  #
  # clients - The Client or Array of Clients to be added.
  #
  # Returns the Bank itself, useful for chaining.
  # Raises Bank::InvalidStateError if the Bank is already closed.
  def add_to_line clients
    unless open?
      raise InvalidStateError.new 'You tried to add clients to an closed Bank'
    end

    self.clients.push(*clients)
    self
  end

  # Public: Count how many Clients waited more then a given time.
  #
  # time - The Integer threshold.
  #
  # Returns the Integer quantity of Clients that waited more then the given
  #   time.
  # Raises Bank::InvalidStateError if the Bank is open.
  def clients_that_waited_more_than time
    if open?
      raise InvalidStateError.new 'You tried to count clients in an open Bank'
    end

    clients.count { |client| client.waiting_duration > time }
  end

  protected

  # Internal: Gets/Sets the Integer representing the clock on the wall.
  attr_accessor :time

  # Internal: Gets/Sets the Integer representing the quantity of available
  # tellers.
  attr_accessor :available_tellers

  # Internal: Calculate how long Clients wait in line. The information is
  # stored in Clients themselves.
  #
  # Returns nothing.
  def calculate_wating_times
    self.time              = 0
    self.available_tellers = tellers

    serve_clients and wait_for_time_to_pass until clients.all?(&:served?)
  end

  def serve_clients
    clients_arrive
    say_hello_to_clients
    perform_tasks_asked_by_clients
    while_other_clients_wait_in_line
    say_goodbye_to_clients
  end

  def clients_arrive
    clients.select(&:before_line?).each do |client|
      client.state = :in_line if client.arrival == time
    end
  end

  def say_hello_to_clients
    while available_tellers > 0 && (next_client = clients.find(&:in_line?))
      allocate_teller
      next_client.state = :being_served
    end
  end

  def while_other_clients_wait_in_line
    clients.select(&:in_line?).each do |client|
      client.waiting_duration += 1
    end
  end

  def perform_tasks_asked_by_clients
    clients.select(&:being_served?).each do |client|
      client.serving_duration += 1
    end
  end

  def say_goodbye_to_clients
    clients.select(&:finished_serving?).each do |client|
      free_teller
      client.state = :served
    end
  end

  def wait_for_time_to_pass
    self.time += 1
  end

  def allocate_teller
    self.available_tellers -= 1
  end

  def free_teller
    self.available_tellers += 1
  end
end

File: bank/c/src/bank.c
/*
 * bank.c
 *
 * Solution to the Bank problem.
 *
 * Refer to `README.md` for description of problem.
 *
 * The solution is all in a single file to work in OBI's auto corrector.
 *
 * The logic for the solution is naively simulating the Bank with the passing
 * of time. Clients are tracked while moved to line and served. Afterwords,
 * the amount of unsatisfied Clients is measured.
 *
 * This approach is simple to write and read, also it's not expensive
 * computation with the limits given in the problem.
 */

#include <stdio.h>
#include <stdlib.h>


/*
 * The maximum time a client should wait in line.
 */
#define MAXIMUM_WAITING_TIME 20

/*
 * Client being served by bank.
 */
typedef struct {
  /*
   * Time when the Client arrives at the bank.
   */
  int arrival;

  /*
   * How long it takes to serve a Client.
   */
  int total_serving_duration;

  /*
   * How long the Client has waited in line.
   */
  int waiting_duration;

  /*
   * How long the teller has been serving the Client.
   */
  int serving_duration;
} client_t;

/*
 * An array of Clients.
 */
typedef struct {
  /*
   * The Clients in the array.
   */
  client_t** clients;

  /*
   * The size of the array of Clients.
   */
  int size;
} clients_t;

/*
 * The Bank that serves the clients.
 */
typedef struct {
  /*
   * Quantity of tellers working in the Bank.
   */
  int tellers;

  /*
   * The amount of tellers busy serving clients.
   *
   * This number should never be greater than `tellers`.
   */
  int occupied_tellers;

  /*
   * Clients outside the Bank, before getting in line.
   */
  clients_t* outside;

  /*
   * Clients in line waiting to be served.
   */
  clients_t* line;

  /*
   * Clients being served.
   */
  clients_t* serving;

  /*
   * Clients that have finished being served.
   */
  clients_t* served;
} bank_t;

/*
 * Read Bank information from `stdin`.
 *
 * Returns the Bank with proper information.
 */
bank_t* read_bank_information();

/*
 * Serve Clients in Bank.
 *
 * bank - The Bank with clients to serve.
 */
void serve_clients(bank_t* bank);

/*
 * Count clients that waited in line more than a threshold.
 *
 * clients   - The line of Clients.
 * threshold - The maximum amount of time a Client should wait in line.
 *
 * Returns the amount of Clients that waited more than the `threshold`.
 */
int clients_exceeded_waiting(clients_t* clients, int threshold);

/*
 * Add Client to array.
 *
 * Addition is performed in the end. This way, if the array is traversed
 * is ascending order, it behaves as a queue.
 *
 * clients - The Clients array in which the Client is added.
 * client  - The Client to be added.
 */
void add_client(clients_t* clients, client_t* client);

/*
 * Remove Client from array.
 *
 * clients  - The Clients array from which the Client is removed.
 * position - The position of the Client being removed. The position is
 *            zero-indexed.
 */
void remove_client(clients_t* clients, int position);

/*
 * Check if all Clients were served.
 *
 * This means the Bank can close.
 *
 * bank - The bank in which the test should be performed.
 */
int all_clients_served(bank_t* bank);

/*
 * Initialize the Client struct.
 *
 * arrival                - The time when the Client arrives at the bank.
 * total_serving_duration - The time it takes to serve a Client.
 *
 * Retuns the initialized Client.
 */
client_t* init_client(int arrival, int total_serving_duration);

/*
 * Free the Client struct.
 *
 * client - The Client to be freed.
 */
void free_client(client_t* client);

/*
 * Initialize Clients array.
 *
 * maximum_size - The maximum amount of Clients in array.
 *
 * Retuns the initialized Clients array.
 */
clients_t* init_clients(int maximum_size);

/*
 * Free Clients array.
 *
 * clients - The Clients array to be freed.
 */
void free_clients(clients_t* clients);

/*
 * Initialize the Bank struct.
 *
 * tellers - The amount of tellers working in the Bank.
 * clients - The amount of Clients the Bank is going to serve.
 *
 * Returns the initialized Bank.
 */
bank_t* init_bank(int tellers, int clients);

/*
 * Free Bank.
 *
 * bank - The Bank to be freed.
 */
void free_bank(bank_t* bank);

int main(int argc, char const *argv[]) {
  int unhappy_customers;
  bank_t* bank;

  bank = read_bank_information();
  serve_clients(bank);
  unhappy_customers = clients_exceeded_waiting(bank->served,
                                               MAXIMUM_WAITING_TIME);
  printf("%d", unhappy_customers);
  free_bank(bank);

  return 0;
}

bank_t* read_bank_information() {
  int c, n, t, d, i;
  bank_t* bank;

  fscanf(stdin, "%d %d", &c, &n);
  bank = init_bank(c, n);

  for (i = 0; i < n; ++i) {
    fscanf(stdin, "%d %d", &t, &d);
    add_client(bank->outside, init_client(t, d));
  }

  return bank;
}

void serve_clients(bank_t* bank) {
  int time, i;
  client_t* client;

  time = 0;

  for (time = 0; ! all_clients_served(bank); ++time) {
    for (i = 0; i < bank->outside->size; ++i) {
      client = bank->outside->clients[i];
      if(client->arrival == time) {
        remove_client(bank->outside, i); --i;
        add_client(bank->line, client);
      }
    }

    for (i = 0; i < bank->line->size; ++i) {
      client = bank->line->clients[i];
      if (bank->occupied_tellers < bank->tellers) {
        remove_client(bank->line, i); --i;
        add_client(bank->serving, client);
        ++bank->occupied_tellers;
      }
      else {
        ++client->waiting_duration;
      }
    }

    for (i = 0; i < bank->serving->size; ++i) {
      client = bank->serving->clients[i];
      ++client->serving_duration;
      if (client->serving_duration == client->total_serving_duration) {
        remove_client(bank->serving, i); --i;
        add_client(bank->served, client);
        --bank->occupied_tellers;
      }
    }
  }
}

int clients_exceeded_waiting(clients_t* clients, int threshold) {
  int i;
  int count;

  count = 0;
  for (i = 0; i < clients->size; ++i) {
    if (clients->clients[i]->waiting_duration > threshold) {
      ++count;
    }
  }

  return count;
}

void add_client(clients_t* clients, client_t* client) {
  clients->clients[clients->size] = client;
  ++clients->size;
}

void remove_client(clients_t* clients, int position) {
  int i;

  --clients->size;
  for (i = position; i < clients->size; ++i) {
    clients->clients[i] = clients->clients[i + 1];
  }
}

int all_clients_served(bank_t* bank) {
  return bank->outside->size +
         bank->line->size +
         bank->serving->size == 0;
}

client_t* init_client(int arrival, int total_serving_duration) {
  client_t* client;

  client = (client_t*) malloc(sizeof(client));
  client->arrival = arrival;
  client->total_serving_duration = total_serving_duration;
  client->waiting_duration = 0;
  client->serving_duration = 0;

  return client;
}

void free_client(client_t* client) {
  free(client);
}

clients_t* init_clients(int maximum_size) {
  clients_t* clients;

  clients = (clients_t*) malloc(sizeof(clients_t));
  clients->clients = (client_t**) calloc(maximum_size, sizeof(client_t*));
  clients->size = 0;

  return clients;
}

void free_clients(clients_t* clients) {
  int i;

  for (i = 0; i < clients->size; ++i) {
    free_client(clients->clients[i]);
  }

  free(clients);
}

bank_t* init_bank(int tellers, int clients) {
  bank_t* bank;

  bank = (bank_t*) malloc(sizeof(bank_t));
  bank->tellers = tellers;
  bank->occupied_tellers = 0;
  bank->outside = init_clients(clients);
  bank->line = init_clients(clients);
  bank->serving = init_clients(clients);
  bank->served = init_clients(clients);

  return bank;
}

void free_bank(bank_t* bank) {
  free_clients(bank->outside);
  free_clients(bank->line);
  free_clients(bank->serving);
  free_clients(bank->served);
  free(bank);
}

